File: problems-originals/0101-0200/0136_single-number/solution_hash.py
from typing import List, Optional


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        # Parity hash set: the first sight of a value adds it, the second
        # removes it — a paired element erases its own trace, so the set
        # holds exactly the values seen an odd number of times.
        seen = set()
        for value in nums:
            if value in seen:
                seen.remove(value)
            else:
                seen.add(value)
        # Fold the odd-count survivors with XOR: even-count values cancel
        # in any XOR fold anyway, so this equals folding the whole array.
        result = 0
        for value in seen:
            result ^= value
        return result

File: problems-originals/0101-0200/0136_single-number/solution_hash.java
import java.util.HashSet;
import java.util.Set;

class Solution {

    public int singleNumber(int[] nums) {
        // Parity hash set: the first sight of a value adds it, the second
        // removes it — a paired element erases its own trace, so the set
        // holds exactly the values seen an odd number of times.
        Set<Integer> seen = new HashSet<>();
        for (int value : nums) {
            if (!seen.remove(value)) {
                seen.add(value);
            }
        }
        // Fold the odd-count survivors with XOR: even-count values cancel
        // in any XOR fold anyway, so this equals folding the whole array.
        int result = 0;
        for (int value : seen) {
            result ^= value;
        }
        return result;
    }
}
